OpenRA.Mods.Common/AI/States/NavyStates.cs

```diff
@@ -0,0 +1,194 @@
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.AI
+{
+	abstract class NavyStateBase : StateBase
+	{
+		protected virtual bool ShouldFlee(Squad owner)
+		{
+			return base.ShouldFlee(owner, enemies => !AttackOrFleeFuzzy.Default.CanAttack(owner.Units, enemies));
+		}
+
+		protected Actor FindClosestEnemy(Squad owner)
+		{
+			if (!owner.IsValid)
+				return null;
+
+			// For ships, cheat and move towards enemy naval production, if any.
+			var t = owner.World.ActorsHavingTrait<Building>().Where(a
+				=> owner.Bot.Info.BuildingCommonNames.NavalProduction.Contains(a.Info.Name)
+				&& a.AppearsHostileTo(owner.Units.First())).FirstOrDefault();
+
+			// If naval yard is too far away, return it.
+			// Else, FindClosest below will find suitable enemy targets :)
+			if (t != null && (t.Location - owner.Units.First().Location).LengthSquared
+					> owner.Bot.Info.MaxBaseRadius * owner.Bot.Info.MaxBaseRadius)
+				return t;
+
+			return owner.Bot.FindClosestEnemy(owner.Units.FirstOrDefault().CenterPosition);
+		}
+	}
+
+	class NavyUnitsIdleState : NavyStateBase, IState
+	{
+		public void Activate(Squad owner) { }
+
+		public void Tick(Squad owner)
+		{
+			if (!owner.IsValid)
+				return;
+
+			if (!owner.IsTargetValid)
+			{
+				var t = FindClosestEnemy(owner);
+				if (t == null) return;
+				owner.TargetActor = t;
+			}
+
+			var enemyUnits = owner.World.FindActorsInCircle(owner.TargetActor.CenterPosition, WDist.FromCells(10))
+				.Where(unit => owner.Bot.Player.Stances[unit.Owner] == Stance.Enemy).ToList();
+
+			if (enemyUnits.Any())
+			{
+				if (AttackOrFleeFuzzy.Default.CanAttack(owner.Units, enemyUnits))
+				{
+					foreach (var u in owner.Units)
+						owner.Bot.QueueOrder(new Order("AttackMove", u, false) { TargetLocation = owner.TargetActor.Location });
+
+					// We have gathered sufficient units. Attack the nearest enemy unit.
+					owner.FuzzyStateMachine.ChangeState(owner, new NavyUnitsAttackMoveState(), true);
+					return;
+				}
+				else
+					owner.FuzzyStateMachine.ChangeState(owner, new NavyUnitsFleeState(), true);
+			}
+		}
+
+		public void Deactivate(Squad owner) { }
+	}
+
+	class NavyUnitsAttackMoveState : NavyStateBase, IState
+	{
+		public void Activate(Squad owner) { }
+
+		public void Tick(Squad owner)
+		{
+			if (!owner.IsValid)
+				return;
+
+			if (!owner.IsTargetValid)
+			{
+				var t = FindClosestEnemy(owner);
+				if (t != null)
+					owner.TargetActor = t;
+				else
+				{
+					owner.FuzzyStateMachine.ChangeState(owner, new NavyUnitsFleeState(), true);
+					return;
+				}
+			}
+
+			var leader = owner.Units.ClosestTo(owner.TargetActor.CenterPosition);
+			if (leader == null)
+				return;
+			var ownUnits = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(owner.Units.Count) / 3)
```

Comment:
Indeed, I copied it. The code is to pause a group of units when the group is spread thin when the units have different movement speeds.  Seems to work in most cases.